### ISSUE TITLE
sol3: topic-conditioned Perplexity allowlist (source librarian proposes, Python admits)

### ISSUE THREAD

**Issue by RichardHightower** [RichardHightower/reliable-agentic-lab#304]
Follow-on to the #285 source wall. The paper loop takes `--topic` on **any** subject (medicine, climate, policy, CS). The current Perplexity seed is a workshop leftover: LangChain / Claude / OpenAI / Stripe docs. A paper that is not about those vendors retrieves almost nothing.

Perplexity accepts at most **20** `search_domain_filter` entries, allow **or** deny, not both. TLD filters are real (`.gov`, not `*.gov`). A fixed 20-host vendor list cannot cover “any topic.” Expanding it by hand with `*.org`, CNN, or [Wikipedia’s list of scientific journals](https://en.wikipedia.org/wiki/List_of_scientific_journals) does not work: that page is titles, not domains, and is thousands of journals across every field. Scientific coverage collapses to a handful of **publisher hosts** (Nature, Science, Elsevier, Springer, Wiley, ACM, IEEE, arXiv), which still cannot all live in a static 20 alongside civic and news sources.

The proposed shape: **one cheap LLM turn proposes up to 20 hosts for this topic; Python admits them.** The comment already in `source_policy.py` stays true: *asking a model to decide what is reputable is how blogs re-enter the paper.* So the model never is the wall.

Saturday Lab 3 stays untouched. No shared library. Copy, do not import. No role holds `Bash`. Writer still does not search.

## What is already there

Both ports own a copy of `source_policy.py` with `SEED_ALLOWLIST` (14 hosts on SDK including `modelcontextprotocol.io` and `sre.google`; 12 on Deep Agents). Python post-filters every URL. GitHub is org-scoped. Deep Agents also has a denylist (`medium.com`, `reddit.com`, `twitter.com`, `x.com`, `pinterest.com`, `quora.com`, `deepwiki.com`, `substack.com`).

Scout is **not a flag**. It is hardcoded inside `PerplexityBackend.search()`: a first Search API pass, then `merge_scout_domains` / `merge_allowlist` may add only `docs.*` / `reference.*` / `learn.*` / known GitHub orgs into leftover slots under the 20 cap. The Agent SDK live path (`task e2e-live` / `--backend agent`) does **not** run Python scout; the researcher subagent issues one MCP `perplexity_search` with the seed. Deep Agents `stage_search` runs scout **inside** the researcher’s one `search` tool call.

Corpus search (#285) sits in front of Perplexity. It does not replace it. Researcher prompt: corpus first, live search for questions the corpus did not answer (prompt, not a Python gate). Verifier: SDK `sections.py` skips `turns.verify()` only when `origin == corpus` **and** `epistemic == corroborated` (two distinct source hashes). `source_supported` still goes live. Deep Agents verify has no corpus short-circuit.

## Why not just enlarge the seed

| Proposal | Verdict |
| --- | --- |
| `arxiv.org` | Yes, as a named host, not via `.org`. |
| `.gov` TLD | Yes. One slot. NIST / NIH / CDC / CISA. |
| `.edu` TLD | Yes, narrowly. Course notes will leak; sourced-row + verify still apply. Scout must **not** promote every `cs.*.edu` into the 20. |
| `.org` TLD | **No.** arXiv / ACM / USENIX / IEEE are `.org`. So are content mills, vendor marketing, advocacy shops, SEO “institutes.” Name the orgs. |
| `cnn.com` as news prototype | **No.** Cable news outranks journals on timely queries. Wire services (Reuters, AP, BBC) are the news-shaped analog of a primary source. |
| `en.wikipedia.org` as a cite | **No.** Use Wikipedia to *design* publisher maps. Do not let the paper cite it. |
| InfoWorld / IEEE Spectrum / ACM Queue | Trade press, last in the existing primary-source order. Named `trade_press` only. |
| Dumping the CS journal list or the multidisciplinary journal list into the filter | Wrong unit. Titles ≠ domains. Publisher hosts cover every field on those pages. |

A static “scholarly-web” seed (`.gov` `.edu` `.int` + Nature/Science/JSTOR + Elsevier/Springer/Wiley + ACM/IEEE + wire services) is a fallback, not the design. The paper can be on any topic; the 20 should be **this topic’s** 20.

## The phase

Insert a **source-allowlist** turn after outline (facets are known) and before any live Perplexity call. One call for the whole run. Cache as `corpus/source_allowlist.json`.

### New reader role: `source_librarian`

- Holds **no** write tool, **no** Perplexity, **no** `Bash`. Optional: read the corpus pack so it can reuse hosts the brain already named. It must not search live (“search to decide where to search”).
- Do **not** give this job to the researcher. A researcher that picks its own allowlist confirms what it wants to say.
- Cheap model is enough.

**Input to the turn:**

- topic + outline section titles
- org-type **enum** (the model may not invent a type):
  `government`, `university`, `standards_body`, `peer_reviewed_publisher`, `preprint`, `professional_society`, `vendor_docs`, `wire_service`, `trade_press`
- hard denylist (see below)
- cap 20
- primary-source order already in the researcher card: spec, official docs, paper, vendor repo, standard, then trade press

**Output schema:**

```json
{
  "domains": [
    {"host": "arxiv.org", "org_type": "preprint", "why": "CS and biomed preprints"}
  ]
}
```

`cable_news` and `encyclopedia` are **not** in the enum. CNN and Wikipedia stay out unless a later issue adds those types on purpose.

### Python is the admission wall

`source_policy.admit(proposed) -> tuple[str, ...]`:

1. Parse. Drop unknown keys. Drop `org_type` not in the enum (`blog` dies here).
2. Normalize host: no scheme, no path, no `*.` wildcards. TLD entries only if they are `.gov` / `.edu` / `.int`.
3. Drop denylist (and any host under a denied host).
4. Cap 20. Stable sort (enum priority then host) so the same outline does not shuffle.
5. Write `corpus/source_allowlist.json` (proposed, dropped, admitted, reasons). Every Perplexity call **and** the post-filter use **this run’s list**, not `SEED_ALLOWLIST`.
6. Existing scout may still add `docs.*` / `reference.*` / `learn.*` / known GitHub orgs into **leftover** slots. Scout cannot add CNN (`news_outlet` is not an enum value).

Denylist (Python, both ports; union with today’s DA list):

`medium.com`, `reddit.com`, `twitter.com`, `x.com`, `pinterest.com`, `quora.com`, `deepwiki.com`, `substack.com`, `en.wikipedia.org`, `wikipedia.org`.

`.org` as a TLD is never admitted. Named `.org` hosts (`arxiv.org`, `acm.org`, `pnas.org`) are fine if their `org_type` is in the enum.

### Fallback

If the librarian turn fails, times out, or admits fewer than N hosts (say 3), fall back to today’s `SEED_ALLOWLIST` so offline fixture and a dead model still run. Fixture tests must not require a live librarian.

Perplexity stays in **allowlist** mode (not denylist). Option A from the design discussion (flip Perplexity to denylist, longer Python wall) is **out of scope** here.

## Role table

Add `source_librarian` as a reader on both ports. Writer remains the only SDK `yes`. Deep Agents: librarian is not a writer. Researcher and verifier still hold `corpus_search` + filtered Perplexity; they consume the admitted list, they do not edit it.

No `Bash` on the librarian. No `Write` on the librarian. Python writes the artifact.

## Ports

Do the Agent SDK folder first, then copy into Deep Agents. Two PRs. Do not extract a shared `source_policy/` package.

| Folder | Today | This issue |
| --- | --- | --- |
| `solutions/sol3_research_agent_sdk` | static `SEED_ALLOWLIST` + optional Python scout on `--backend perplexity` | librarian turn + `admit()` + run allowlist on MCP **and** Python Perplexity lanes |
| `solutions/sol3_research_deep_agents` | static seed + scout inside `search()` | the copy; `stage_search` uses the run list |

SDK live (`SdkTurns.research` / researcher MCP) must pass the admitted list into the researcher prompt **and** into the MCP `search_domain_filter`. Today the allowlist is pasted as prose in `research-researcher.md`. That paste becomes the run artifact, or the researcher is told to use only hosts in `corpus/source_allowlist.json` and Python still post-filters.

## Tests

- Librarian proposes `medium.com` → dropped.
- Proposes `.gov` with `org_type: government` → kept as TLD `.gov`.
- Proposes `cnn.com` as `wire_service` → dropped (wrong type).
- Proposes `cnn.com` as `trade_press` → dropped (still not in enum as cable news; host is not trade press — pin this).
- Proposes `infoworld.com` as `trade_press` → kept.
- Proposes `en.wikipedia.org` as anything → dropped by denylist.
- Proposes `.org` TLD → dropped.
- Proposes `arxiv.org` as `preprint` → kept.
- 21st host → truncated, stable order.
- Librarian failure → fallback seed, run still proceeds.
- Offline fixture / `task demo` never calls a live librarian (stub or seed fallback).
- Post-filter rejects a URL whose host is not on the admitted list even if Perplexity returned it.
- Scout can add `docs.example.com` into a leftover slot; cannot add `medium.com`.

## Definition of done

- `task test` in the folder is green. Report the count in the PR.
- `python3 loop.py --table-only` still shows `yes` only on the writer in the SDK port; `source_librarian` prints `no`.
- A live or recorded turn writes `corpus/source_allowlist.json` with `proposed`, `dropped`, `admitted`.
- Every Perplexity call in that run uses `admitted` (max 20).
- SPEC.md / SKILL.md stay in lockstep with `test_docs.py`.
- One commit-msg issue ref (`Refs #NNNN` of this issue).
- Two PRs: SDK first, then DA copy.

## Out of scope

- Flipping Perplexity to denylist mode.
- Topic-conditioned *extra* overlays beyond this one turn (e.g. a medical-only PubMed plugin).
- DNS/HTTP liveness checks on proposed hosts (follow-on; stale training data is accepted for v1).
- Adding `cable_news` or `encyclopedia` to the enum.
- Saturday Lab 3.
- Changing corpus-search order, chartist, or ingest-brain.
- A shared source-policy package.

## Implementation notes

- `source_policy.py`: denylist + enum + `admit()`.
- New reader agent/skill + one `turns.source_allowlist(topic, outline)`.
- LINEAR / STAGE_ORDER: after outline, before research/search. Skip when the artifact already exists (`--resume`).
- Keep today’s seed as **fallback only**.
- Scout rules unchanged; they fill leftover slots, they do not replace `admit()`.
